Async section - Overview, C#/VB, and F#

cartermp opened the pull request:
I feel that async programming is such a core concept in modern and future .NET dev that it demands its own section (see #24) that can give people the minimum necessary information to get started with async programming.  This includes a short overview (what is it, why do you need it), and language-specific articles that discuss what it is from the programmer's perspective without implementation details, a useful code sample for each language, and some important advice/gotchas to know.

New section is formatted as such:

Async Programming
|
|__ Async with .NET Overview
|
|__ C# and VB async programming
|
|__ F# async programming

cartermp committed:
Rough draft of about half of the async stuff

cartermp committed:
Fix typo

cartermp committed:
Fix formatting issue in code sample

cartermp committed:
Add info/advice

cartermp committed:
Added inline async handler snippet and various fixes

I felt like adding an inline event handler that's async would be helpful
for those writing Xamarin apps.

cartermp committed:
Final adjustments before review request

Removed the "what is async actually doing" because that isn't relevant
for someone who just wants to start using it.  Reworded the first couple
of bullet points to try and be more clear.  Various tiny formatting
fixes.

cartermp committed:
Adding myself as an author

cartermp committed:
Adding VB code sample and splitting off general stuff into the overview

cartermp committed:
Adding F# page and some minor formatting.

cartermp committed:
Function name change

cartermp committed:
Adding similarities and differences between C# and F# async

cartermp committed:
Updates!

Updated the F# code sample to be something which actually compiles,
clarified some points about differences between C#/VB async, and some
minor formatting.

cartermp committed:
Minor fixes

Renamed files for easier relative linking, clarified some points, and
typo fixing.

cartermp commented:
@NumberByColors is going to be doing some significant work on the F# article, so I'd expect its contents to change a lot.

cartermp commented:
/cc @richlander @terrajobst @blackdwarf for review

blackdwarf commented on docs/async/async-csharp-vb.rst:
```diff
@@ -0,0 +1,167 @@
+
+Although async programming is relatively straightfoward, there are some details to keep in mind which could otherwise result in some nasty behavior.
+
+* Do yourself a favor and append "Async" to the end of every async method you write which could be consumed by another method.
```
Maybe formulate it more as a best practice advice? For instance "Best practice is to add an "Async" suffix..." etc.

blackdwarf commented on docs/async/async-csharp-vb.rst:
```diff
@@ -0,0 +1,167 @@
+
+A few important things to know before continuing:
+
+* ``Task<T>`` and ``Task`` are used a lot in async code.  These are abstractions that can be thought of as "something that needs to get done and could be long-running".
```
I would maybe put in some more explanation on what are Task<T> and Task, given that they are so essential to the entire feature.

blackdwarf commented on docs/async/async-csharp-vb.rst:
```diff
@@ -0,0 +1,167 @@
+
+* Do yourself a favor and append "Async" to the end of every async method you write which could be consumed by another method.
+
+Yes, it's sort of hungarian notation which is widely hated, but being extra explicit is a lot better than tracking down a race condition.  Note that certain methods which aren't explicity called by your code (such as event handlers or web controller methods) may not necessarily apply.
```
I wouldn't know that Hungarian notation is widely hated. :) Would rephrase this.

cartermp replied:
Agreed!

blackdwarf commented on docs/async/async-csharp-vb.rst:
```diff
@@ -0,0 +1,167 @@
+
+Failing to apply the ``await`` operator to a task will result in the async method running synchronously!  Application of ``await`` is what suspends the async method, giving back control to the method which called it.  Pay attention to compiler warnings about this.
+
+* ``async void`` should only be used for event handlers.
```
Is this something that is a technical requirement (i.e. a compiler will throw an error or a warning) or a best practice? I think whichever it is should be spelled out.

blackdwarf commented on docs/async/async-overview.rst:
```diff
@@ -0,0 +1,30 @@
+Here are a few reasons:
+
+* Responsiveness in apps and services is expected by users, and even slight UX hangups are often punished harshly (one-star reviews).
+* Almost all modern apps demand elements which block on I/O in some way.
```
This feels incomplete. I would add a line that ties it into the first bullet, that is, the point is that because modern apps demand elements that block on I/O, it is easy to create a non-responsive UX, or aysnc is needed but usually complex.

blackdwarf commented on docs/async/async-overview.rst:
```diff
@@ -0,0 +1,30 @@
+
+.NET provides the capability for apps and services to be responsive with easy-to-use, language-level asynchronous programming models in C#/VB and F#.  Although similar, the nuances of async programming in F# differs from that in C# and VB.
+
+Why Async?
```
Why did we implement it or why it is needed at all? I think it would read better if the title is something along the lines of "Why intrinsic support for async?" or something similar.

cartermp replied:
How about "Why Write Async Code?"?

blackdwarf replied:
Sounds good.

blackdwarf commented on docs/async/async-csharp-vb.rst:
```diff
@@ -0,0 +1,167 @@
+* Marking a method as ``async`` tells the compiler that you're expecting to ``await`` some blocking task.
+* ``await`` can only be used inside an async method.
+* When the ``await`` keyword is applied, it suspends the async method and returns control back to its caller until the awaited task is complete.  This is what makes it responsive.
+* Unless an async method has an ``await`` inside its body, it will run synchronously, defeating the whole point of making it async!
```
Nitpicking: leave out the "defeating the whole point" and just leave the fact that async methods will be sync unless await is used.

blackdwarf commented on docs/async/async-csharp-vb.rst:
```diff
@@ -0,0 +1,167 @@
+	
+That said, ``async void`` is perfect for event handlers where the event involves any blocking task(s).
+
+* Avoid async lambda expressions when combined with other async code
```
Looking at this, I can read it as "avoid async lambdas", but then below you mention lambdas only in the context of LINQ (which is a good point). If this is only about LINQ, make it clearer on this line.

blackdwarf commented:
Like the content, left some notes. One thing I just realized on this PR is that we should have a guidance  on how to deal with lines in the text so that we can leave good notes.

cartermp commented:
I feel like elaborating on the `Task<T>` concept would be better suited if it were in a separate article altogether.  Should have be part of this, or should that be a separate branch?  I can see the value in making it a part of this branch mainly because `Task<T>` is used primarily with `async` and `await` anyways.

cartermp committed:
Addressing comments from Z

* Made it clear that async lambdas were bad in the context of LINQ
* Addressed "defeating the whole point" nitpick
* Changed "Why Async?" title to "Why Write Async Code?"
* Combined bullet points about responsiveness and modern apps which
block on I/O
* Removed Hungarian Notation bit
* Hinted at Task<T> being a more complex thing that deserves elaboration
* Added links to C#/VB and F# pages in the async overview

blackdwarf commented:
For sure, I would put it in this document. It is not overly long, and devs who use async need to know about Task<T> and Task to be able to full it grasp the entire "machinery".

cartermp commented:
Mmkay, will do.

cartermp committed:
Adding a "more on tasks" section

As per Z's suggestion, this section talks a bit about Task and Task<T>
from the perspective of the async model.  It also gives links to TPL
data and task-parallel documentation, should the user be interested in
using tasks for more than just coordinating async code.

cartermp committed:
Adding new advice bullet

Added suggestion to write less stateful code.  Re-worded the section on
coordinating blocking code.  Fixed a few typos as well.

cartermp commented:
Something I realized, especially after speaking with some past coworkers who do .NET code, is that it's not entirely clear what the relationship between `Web Server <---> Async Code` is, and why IIS or other technology will not be enough to handle the needs of a service to scale vertically.  I feel like the overview section could expand on this.

Thoughts?

blackdwarf commented:
That is an interesting notion. Does it belong in the primer on async though? I guess it could if you keep it very short and just expand on some of the points that you already made. I wouldn't want the Primer, being imagined as something a dev new to .NET can read in a couple of hours and get some overview, to become too large and in-depth. That would defeat its purpose, methinks.

cartermp commented:
Hmmm.  Well I wouldn't put it in the primer, but now that I think about it, the async overview article may be a little too much to stick it in.  Basically what I'm envisioning is an article that describes exactly what's going on in an async context.  Specifically,

- How it leverages the OS to perform the work rather than spawning off a new thread
- A dive into what's actually happening on the device side of things when making a network call
- A dive into what's actually happening on the server side of things when a request needs to extract data from a database

The server side would talk about how it also works from the perspective of IIS, so I can really get the point across that IIS being multithreaded only helps web servers so much.

This would involve a few diagrams as well, and would be linked to from the articles I've written and probably the Primer article.

I've already sort of done this, but just in-person.  I gave a talk at my old workplace about why they should adopt async programming into their workflow (especially from a student-preparedness level).  This would basically be the more refined version of that talk.

cartermp committed:
Clarifying server load bullet and task section

Clarified that Tasks in async programming aren't threads at all, alluded
to a future article about what's really going on, and made it clear that
the role of async programming in a server environment is to scale a
system upwards.

blackdwarf commented:
Well, RTD, at this point, is more for intro docs for the actual platform. I would like to see a more in-depth article, but you can always put it on a blog somewhere and then, once/if we figure out where it needs to go, we take it from there. :)

cartermp commented:
Will do.  Putting that stuff on hold for now.

cartermp committed:
Clarity and some fixes

* Addressed some typos.
* Made clear in the overview that this is for block I/O, and in a server
environment aims to aid in scaling up
*

cartermp committed:
Code sample rename and functional reword

Small change, but I wanted to do the following:

* Rename code sample service calls to be more realistic
* Point readers to the specific tenant of functional programming that
they should aim for, rather than just say "be more functional"

cartermp committed:
Clarity and Primer Change

* Changed the async section in the primer to be smaller and link to the
async overview article.
* Changed C#/VB samples to be more accurate and have better comments for
the inline sample
* Made the last advice piece more clear about goals to achieve

richlander commented:
On the web server / async question, async does help that scenario a lot. Every request hitting the server needs a thread. Threads require machine resources (memory), so there is a limited number that a given machine can have. Given enough requests, the machine can keel over. What async does is free up threads for use on other requests while (a)waiting on I/o operations.

See this blog post: http://blogs.msdn.com/b/dotnet/archive/2012/11/13/the-net-framework-4-5-is-optimized-for-the-cloud.aspx

richlander commented on docs/async/async-csharp-vb.rst:
```diff
@@ -0,0 +1,202 @@
+==========================
+By `Phillip Carter`_
+
+From a developer's perspective, the C#/VB async model is some nice syntax that hides complexity and allows you to focus on writing methods without worrying about things like thread pools and callbacks.
```
You are missing the value-prop. While it's true that it is nice syntax that hides complexity, it doesn't state what the end goal is. What does it all achieve?

richlander commented on docs/async/async-csharp-vb.rst:
```diff
@@ -0,0 +1,202 @@
+
+From a developer's perspective, the C#/VB async model is some nice syntax that hides complexity and allows you to focus on writing methods without worrying about things like thread pools and callbacks.
+
+The core of this are the ``async`` and ``await`` keywords (``Async`` and ``Await`` in VB.NET), which you sprinkle over otherwise normal methods. 
```
I think it's just "VB", not "VB.NET".

It's true that the methods are otherwise normal. That said, there are a few categories of good candidates for async/await. I/O bound is one and the other is work you want to get off the main thread (do in the background). There are some more niche scenarios, too.

cartermp replied:
I'm going to re-word this to state that it allows you to express your intent very easily and allows the code to remain simple.

To your point about candidates for async code, I think that should be an addition to the Overview document.  Although C#/VB and F# differ quite a bit in how async code is handled, they are fundamentally used for the same stuff (e.g. you wouldn't use async code to try and solve a CPU-bound task).

richlander commented on docs/async/async-csharp-vb.rst:
```diff
@@ -0,0 +1,202 @@
+
+A few important things to know before continuing:
+
+* ``Task<T>`` and ``Task`` are used to represent the return types for async code.  These are abstractions that can be thought of as "something that needs to get done and could be long-running".  `More on Task and Task<T>`_.
```
I'd describe this slightly differently. I'd start with describing what Task represents and then state that it's the return type of async methods. One follows the other. There is no affinity to Task and long-running code. When I read long-running code, I think CPU-bound. A big scenario of async is I/O-bound code.

richlander commented on docs/async/async-csharp-vb.rst:
```diff
@@ -0,0 +1,202 @@
+A few important things to know before continuing:
+
+* ``Task<T>`` and ``Task`` are used to represent the return types for async code.  These are abstractions that can be thought of as "something that needs to get done and could be long-running".  `More on Task and Task<T>`_.
+* Marking a method as ``async`` tells the compiler that you're expecting to ``await`` some blocking task.
```
I think "blocking" is the wrong term. By virtue of the code being awaitable, it's not blocking. Recall that you don't have to await the code at that point.

richlander commented on docs/async/async-csharp-vb.rst:
```diff
@@ -0,0 +1,202 @@
+* ``Task<T>`` and ``Task`` are used to represent the return types for async code.  These are abstractions that can be thought of as "something that needs to get done and could be long-running".  `More on Task and Task<T>`_.
+* Marking a method as ``async`` tells the compiler that you're expecting to ``await`` some blocking task.
+* ``await`` can only be used inside an async method.
+* When the ``await`` keyword is applied, it suspends the async method and returns control back to its caller until the awaited task is complete.  This is what makes it responsive.
```
Make sure it is clear which async method is suspended. There are multiple async methods at play.

blackdwarf commented:
That is completely true. However, the question is not whether it is true, but whether this is the place for this kind of story. This document, that is. :)

richlander commented on docs/async/async-csharp-vb.rst:
```diff
@@ -0,0 +1,202 @@
+* Marking a method as ``async`` tells the compiler that you're expecting to ``await`` some blocking task.
+* ``await`` can only be used inside an async method.
+* When the ``await`` keyword is applied, it suspends the async method and returns control back to its caller until the awaited task is complete.  This is what makes it responsive.
+* Unless an async method has an ``await`` inside its body, it will run synchronously!
```
To my mind, it's not so much that it will run 'synchronously', but that it won't yield. It could be the case that the methods that it calls are still handled asynchronously.

richlander commented on docs/async/async-csharp-vb.rst:
```diff
@@ -0,0 +1,202 @@
+* ``await`` can only be used inside an async method.
+* When the ``await`` keyword is applied, it suspends the async method and returns control back to its caller until the awaited task is complete.  This is what makes it responsive.
+* Unless an async method has an ``await`` inside its body, it will run synchronously!
+* Return types for async methods should always be either ``Task<T>`` or ``Task``, with the exception of event handlers (these are ``void``).  More on this later.
```
I'd probably avoid the use of the word "exception", given that the mind quickly jumps to another topic.

You could say "required to return void" for event handlers. You just described the return types at the top of the list. Any reason you need to do it again?

richlander commented on docs/async/async-csharp-vb.rst:
```diff
@@ -0,0 +1,202 @@
+
+	private async void GetDirectionsForPickup_Pressed(object sender, CustomUserDataEventArgs e)
+	{
+	    var user = e.Source as BasicUserData;
```
I think that this example is too complicated for a simple example. I propose a simple HttpClient call and deserialization of JSON data, and may updating a label.

cartermp replied:
My thought process was the exact opposite - typically whenever I view docs there's something too simple and won't ever resemble actual problems I'm trying to solve.

Do we have any data on what people generally prefer?  If it turns out that people do generally want something like 2-3 lines of code, we should do that.  I want to make sure that people don't look at these docs and say, "great, but that doesn't resemble my problem".

blackdwarf replied:
I generally agree with Philip, the abstract examples are pretty boring (which is why I **severly dislike** all examples of recursion out there), but there is a middle ground, that is, something that is useful but is not so tied to a specific knowledge around a technology (like it is here). We could do a server listening on a port, a more complex example of getting JSON data and parsing it and doing something etc.

cartermp replied:
How about keeping a larger sample such as this (maybe even keep this?), but moving a super-simple sample up near the top?  I think there's value in seeing code immediately in the intro, and keeping it down to 2-3 lines is a good candidate.

richlander replied:
@blackdwarf On recursion, do you need an example more complicated than Fibonacci? It's super short, but clear on the mechanism. No?

I'm not suggesting a more 'abstract' example, just one less specific. The real issue is that directly using FX APIs makes them more familiar to the reader. The example uses user APIs, which means you have to guess at what they do as part of understanding the async mechanism.

cartermp commented:
I'm torn on the idea of having a more advanced document describing what's going on.

One one hand (as evidenced by the Hacker News Thread), there seem to be quite a few people who don't know what's actually going on with async code.  I used to be too, and there really isn't much existing literature on the topic. (FYI I'm the one who responded to the OP on the HN thread).

On the other hand, I agree with Z that it may be too in-depth of a topic for these docs, especially if they're being used as a "get familiar with .NET so you can start coding" mechanism.

cartermp committed:
Updates based on feedback from Rich

richlander commented on docs/async/async-csharp-vb.rst:
```diff
@@ -0,0 +1,200 @@
+		
+	    // This independent work can be done concurrently
+	    // since it doesn't rely on directions!
+	    NotifyUserOfPickupAsync(user.Contact);
```
This line doesn't seem right. I would expect you want to assign this to a variable of Task<T> and then await it later. Otherwise, it won't be concurrent. I also don't expect that this code will compile as-is.

richlander commented on docs/async/async-csharp-vb.rst:
```diff
@@ -0,0 +1,200 @@
+	    return json;
+	}
+	
+Bigger Example (VB)
```
We should also add F# examples. I see you have a separate doc for F#. We should add a link to it.

richlander commented on docs/async/async-csharp-vb.rst:
```diff
@@ -0,0 +1,200 @@
+==========================
+By `Phillip Carter`_
+
+Async programming is traditionally associated with callbacks and a general lack of straightforward code.  One might even say that straightforward code and scalable code are often at odds with one another.  This is not a problem with the C#/VB async model!  It is a language-level construct which hides complexity and allows you to focus on the problem at hand without making the code confusing in th ...
```
It seems odd to refer it the model as the "C#/VB async model". Task is not a language thing. The syntax is.

cartermp replied:
> It seems odd to refer it the model as the "C#/VB async model". Task is not a language thing. The syntax is.

I'd argue it's far more than a syntax thing.Even though C#/VB and F# implement the same overall model (Promises), they approach it differently.  The F# approach is based on computation expressions which treat async operations as expressions of work to be done in some asynchronous context.

The C#/VB model much more straightforward conceptually, using objects which represent the operation running right now (`Task` and `Task<T>`) as things you pass around and eventually wait to complete.  F#'s async does not use the same types either; it uses the `Async<'a> type, which although is used roughly the same way does result in things being a little different (namely it makes the compiler even stricter than it is without async workflows).

The least subtle difference in terms of use is that you cannot nest tasks (e.g. `await Foo(await GetResult(), bar);` is impossible in F#).  There are also numerous more subtle (and devious) things you can do with C#/VB which can cause the code to not run asynchronously.  It's literally impossible to have that happen in F# unless you input that async operation into the `Async.RunSynchronously` function.

I'd like to avoid conflating anything F#-related in this document to avoid giving the impression that they're operating similarly.

richlander commented on docs/async/async-csharp-vb.rst:
```diff
@@ -0,0 +1,200 @@
+More on Task and Task<T>
+------------------------
+
+As mentioned before, Tasks are constructs which represent long-running operations calling code must wait on.
```
There is no specific affinity between Tasks and "long-running operations". LROs is one of the use cases.

richlander commented on docs/async/async-csharp-vb.rst:
```diff
@@ -0,0 +1,200 @@
+
+As mentioned before, Tasks are constructs which represent long-running operations calling code must wait on.
+
+* ``Task`` represents a single operation which does not return a value.
```
It could be useful to state that these types are somewhat similar to Func and Action. Maybe that's too meta.

richlander commented on docs/async/async-csharp-vb.rst:
```diff
@@ -0,0 +1,200 @@
+* ``Task`` represents a single operation which does not return a value.
+* ``Task<T>`` represents a single operation which returns a value of type ``T``.
+
+It's important to reason about Tasks as abstractions of work to be done rather than threads, largely because they typically aren't threads at all!  Asynchronous .NET library calls essentially delegate work to the operating system, which is naturally asynchronous when I/O is performed.  Let this be made clear once more: when calling an async .NET library method, there is no new thread spawned to h ...
```
This is kinda odd. Earlier, I read about long running work. There, I think of non I/O work and the Task.Run pattern, which would result in using a new thread. Here, I read about the I/O scenario, which might not be long-running.

richlander commented on docs/async/async-csharp-vb.rst:
```diff
@@ -0,0 +1,200 @@
+
+Async programming is traditionally associated with callbacks and a general lack of straightforward code.  One might even say that straightforward code and scalable code are often at odds with one another.  This is not a problem with the C#/VB async model!  It is a language-level construct which hides complexity and allows you to focus on the problem at hand without making the code confusing in th ...
+
+The core of this are the ``async`` and ``await`` keywords (``Async`` and ``Await`` in VB), which allow for simple changes to code with a huge impact.  For example:
```
Well, the core is really Task. The keywords + Task are the developer experience.

richlander commented on docs/async/async-csharp-vb.rst:
```diff
@@ -0,0 +1,200 @@
+	    DisplayData(data);
+	};
+	
+And that's it!  No callbacks and no library to wrestle with.
```
Any reason to bring up a library? What's that referring to?

richlander commented on docs/async/async-csharp-vb.rst:
```diff
@@ -0,0 +1,200 @@
+
+For those who are more theoretically-inclined, the model used by C# and VB is an implementation of the `Future and Promise concurrency model <https://en.wikipedia.org/wiki/Futures_and_promises>`_.
+
+A few important things to know before continuing:
```
Some of this should probably be explained in more depth.

richlander commented on docs/async/async-csharp-vb.rst:
```diff
@@ -0,0 +1,200 @@
+
+A few important things to know before continuing:
+
+* Async code relies on ``Task<T>`` and ``Task``, which are constructs used to abstract the notion of a job that you need to wait on for a result.  They also act as the return types, except for the case of Event Handlers (these return ``void``).  `More on Task and Task<T>`_.
```
The use of "relies" doesn't quite seem right. I think of it more that async work is modeled as Tasks. The fact that Task is also the return type is fine to mention, but it also stands to reason. I wouldn't mention the "async void" case as part of the definition. It's really something special and not a core part of the model/experience.

richlander commented on docs/async/async-csharp-vb.rst:
```diff
@@ -0,0 +1,200 @@
+A few important things to know before continuing:
+
+* Async code relies on ``Task<T>`` and ``Task``, which are constructs used to abstract the notion of a job that you need to wait on for a result.  They also act as the return types, except for the case of Event Handlers (these return ``void``).  `More on Task and Task<T>`_.
+* Marking a method as ``async`` tells the compiler that you're expecting to ``await`` the completion of a task.   Application of ``await`` is optional, but recommended in most cases.
```
Doesn't 'async' just mean that the method will call async methods and that it is awaitable? It doesn't (in my mind) make a statement about one expects to do. You can just as easily using ContinueWith or block on the Wait method. Right?

Await is recommended for apps. It isn't so much for libraries, unless necessary.

cartermp replied:
I'm just going to get rid of this bullet point.  I already call out the gotcha of an async method without an `await` running synchronously later.

richlander commented on docs/async/async-csharp-vb.rst:
```diff
@@ -0,0 +1,200 @@
+* Async code relies on ``Task<T>`` and ``Task``, which are constructs used to abstract the notion of a job that you need to wait on for a result.  They also act as the return types, except for the case of Event Handlers (these return ``void``).  `More on Task and Task<T>`_.
+* Marking a method as ``async`` tells the compiler that you're expecting to ``await`` the completion of a task.   Application of ``await`` is optional, but recommended in most cases.
+* ``await`` can only be used inside an async method.
+* When the ``await`` keyword is applied, it suspends the calling method and returns control back to its caller until the awaited task is complete.  This is what makes it responsive.
```
"responsive" is a client concept, while async/await is equally applicable to the server.

cartermp replied:
A synonym sometimes used on the server side is "elastic".  I'll use that to avoid any confusion on terminology.

richlander commented on docs/async/async-csharp-vb.rst:
```diff
@@ -0,0 +1,200 @@
+* ``await`` can only be used inside an async method.
+* When the ``await`` keyword is applied, it suspends the calling method and returns control back to its caller until the awaited task is complete.  This is what makes it responsive.
+* Unless an async method has an ``await`` inside its body, it will never yield!
+* Although not strictly required, ``async void`` should **only** be used on Event Handlers.  More on this later.
```
Might be useful to say:

" should only be used with event handlers, where it is required."

richlander commented on docs/async/async-csharp-vb.rst:
```diff
@@ -0,0 +1,200 @@
+Tasks are awaitable, meaning that the application of the ``await`` keyword will allow for either:
+
+* "Unwrapping" the return value for a ``Task<T>`` operation after it completes.
+* Simply waiting for a ``Task`` operation to finish before moving forward.
```
It's waiting w/o blocking.

richlander commented on docs/async/async-csharp-vb.rst:
```diff
@@ -0,0 +1,200 @@
+* "Unwrapping" the return value for a ``Task<T>`` operation after it completes.
+* Simply waiting for a ``Task`` operation to finish before moving forward.
+
+Tasks are also used outside of the async programming model.  They are the foundation of the Task Parallel Library, which allows support for `Data Parallelism <https://msdn.microsoft.com/en-us/library/dd537608(v=vs.110).aspx>`_ and `Task Parallelism <https://msdn.microsoft.com/en-us/library/dd537609(v=vs.110).aspx>`_.  However, from the perspective of writing responsive code with the C#/VB async p ...
```
You might be clear on which async mode (TAP) you mean. There are 3.

richlander commented on docs/async/async-csharp-vb.rst:
```diff
@@ -0,0 +1,200 @@
+Important Info and Advice
+-------------------------
+
+Although async programming is relatively straightfoward, there are some details to keep in mind which could otherwise result in some nasty behavior.
```
Can you use a more descriptive word than "nasty"?

richlander commented on docs/async/async-csharp-vb.rst:
```diff
@@ -0,0 +1,200 @@
+
+Although async programming is relatively straightfoward, there are some details to keep in mind which could otherwise result in some nasty behavior.
+
+* **Best practice is to add "Async" to the end of every async method name you write which could be consumed by another method.**
```
Isn't every method in place so that it will get called by another method?

This is passive construction. How abut "You should add 'Async' as the suffix of every async method."

richlander commented on docs/async/async-csharp-vb.rst:
```diff
@@ -0,0 +1,200 @@
+
+* **Best practice is to add "Async" to the end of every async method name you write which could be consumed by another method.**
+
+Failure to do so could result in having to track down a race condition later.  It's better to be explicit here!  Note that certain methods which aren't explicity called by your code (such as event handlers or web controller methods) may not necessarily apply.
```
Will it result in a race condition?

richlander commented on docs/async/async-csharp-vb.rst:
```diff
@@ -0,0 +1,200 @@
+
+* ``await`` **is what will ultimately make a method asynchronous.**
+
+Failing to apply the ``await`` operator to a task will result in the async method running synchronously!  Application of ``await`` is what suspends the async method, giving back control to the method which called it.  Pay attention to compiler warnings about this.
```
This isn't really exactly true. You can assign a task-returning method call to a Task. Doing so won't block.

How about "yielding control". It might be useful to clarify why that is valuable.

richlander commented on docs/async/async-csharp-vb.rst:
```diff
@@ -0,0 +1,200 @@
+
+* ``async void`` **should only be used for event handlers.**
+
+Why?  That's the only reason they were allowed in the first place!  Async programming uses the ``Task`` and ``Task<T>`` objects, which provide flexibility in coordinating asynchronous work.  Throwing that out of the window with ``async void`` doesn't follow the model very well.  Here's some specific issues:
```
You might clarify that this pattern is required to make the scenario work.

I don't think of the types as "providing flexibility", but representing async work in a first class way.

Try to avoid language like "throwing that out of the windows". It's not clear what you mean.

richlander commented on docs/async/async-csharp-vb.rst:
```diff
@@ -0,0 +1,200 @@
+	
+That said, ``async void`` is perfect for event handlers where the event involves any blocking task(s).
+
+* **Tread carefully when using async lambdas in LINQ expressions**
```
This is begging for a link to a document that provides more detail and guidance.

cartermp replied:
Agreed.

richlander commented on docs/async/async-csharp-vb.rst:
```diff
@@ -0,0 +1,200 @@
+====================== ================================= =======================
+``await``              ``Task.Wait`` or ``Task.Result``  Retreiving the result of a background task
+``await Task.WhenAny`` ``Task.WaitAny``                  Waiting for any task to complete
+``await Task.WhenAll`` ``Task.WaitAll``                  Retreiving the results of multiple tasks
```
Symmetric with the prior lines ... "Waiting for all tasks to complete".

Retreiving? sp?

richlander commented on docs/async/async-csharp-vb.rst:
```diff
@@ -0,0 +1,200 @@
+``await``              ``Task.Wait`` or ``Task.Result``  Retreiving the result of a background task
+``await Task.WhenAny`` ``Task.WaitAny``                  Waiting for any task to complete
+``await Task.WhenAll`` ``Task.WaitAll``                  Retreiving the results of multiple tasks
+``await Task.Delay``   ``Thread.Sleep``                  Waiting for a period of time
```
The before and after makes sense for the first row and not any of the others.

richlander commented on docs/async/async-csharp-vb.rst:
```diff
@@ -0,0 +1,200 @@
+``await Task.Delay``   ``Thread.Sleep``                  Waiting for a period of time
+====================== ================================= =======================
+
+The good news is that calling async code which has no return value has no caveats.  ``JustDoesAJobAsync()`` does not need to be coordinated unless the calling method depends on its execution.  However, this would warrant a refactor, bringing up the final point...
```
I thought that this still style of method call still needed to be assigned to a variable. No?

Why would it need refactoring?

richlander commented on docs/async/async-csharp-vb.rst:
```diff
@@ -0,0 +1,200 @@
+
+* **Write less stateful code**
+
+Depend less on the state of objects and the exeuction of certain methods when writing async code.  Instead, depend on the return values of methods.  Why?
```
exeuction? sp?

I suspect you mean "global objects"?

cartermp replied:
> I suspect you mean "global objects"?

I mean any object that isn't the return type of a method you've called.  Usually this means some global object, but certain things like passing `out` and `ref` parameters into methods *which call* async methods and relying on those parameters can result in control flow that forces the return type of the async operation to be treated differently based on something that happened as a result of wishing to (conceptually, at least) invoke something asynchronously.

Basically, I'm advocating for being completely free of side effects here, because reasoning about async code when you're throwing it together with side effects can make otherwise straightforward code much more challenging to work with.

cartermp committed:
Updates on rlander feedback sans the samples

I'll probably need some more time to come up with good samples so I'll
push these up independently.

cartermp commented:
Made quite a few updates based on @richlander feedback.  I decided to not do anything about the samples yet because feedback on the other stuff can be independent from that.  I'll probably need a bit more time to come up with a decent scenario for a sample which uses only .NET code.

richlander commented on docs/async/async-csharp-vb.rst:
```diff
@@ -0,0 +1,198 @@
+
+Async programming in C# and VB share a language-level asynchronous programming model which allows for easily writing asynchronous code without having to juggle callbacks or conform to a library which supports asynchronicity.
+
+The core of this are the use of ``async`` and ``await`` keywords (``Async`` and ``Await`` in VB) and either direct or implicit use of the ``Task`` and ``Task<T>`` objects.  The result is powerful code that retains simplicity and ease of understanding.
```
"make" or "ly"

cartermp committed:
Addressed richlander comments and Redid samples

The samples here use only .NET FX libraries.

cartermp commented:
Addressed @richlander comments and re-did the C#/VB samples to be only .NET FX code.

cartermp committed:
F# async updates (1/2)

Pushing half of this now so that I can work on these on my laptop.

cartermp committed:
Finished F# section minus large sample and resources

Needs to have a larger sample doing something non-trivial, and a
recommended list of further reading.  I'd prefer to have David or an F#
dev put together the recommended reading list.

cartermp commented:
Fleshed out the F# section after speaking with David.  Should be a "minimum viable product" at this stage, in that it calls out what's important to know about the F# model, some special syntax to know, some samples, and a compare/contrast for C#/VB folks.

cartermp committed:
Completed all but larger sample for F# Async

* Put important links in Further Resources
* Removed Async.StartImmediate
* Added code sample to Async.Start
* Fixed some spelling and inconsistent wording

richlander commented on docs/async/async-csharp-vb.rst:
```diff
@@ -0,0 +1,199 @@
+
+Async programming in C# and VB share a language-level asynchronous programming model which allows for easily writing asynchronous code without having to juggle callbacks or conform to a library which supports asynchronicity.
+
+The core it all are the ``Task`` and ``Task<T>`` objects, which model arbitrary asynchronous operations.  These are interacted with via the ``async`` and ``await`` keywords (``Async`` and ``Await`` in VB) to provide a natural-feeling way to write async code.  The result is the ability to write powerful code which cleanly expresses intent.  There are other ways to approach this than ``async`` and  ...
```
I think you are missing something ... "The core it all are" ??

richlander commented on docs/async/async-csharp-vb.rst:
```diff
@@ -0,0 +1,199 @@
+A few important things to know before continuing:
+
+* Async code uses ``Task<T>`` and ``Task``, which are constructs used to model the work being done in an asynchronous context.  `More on Task and Task<T>`_
+* When the ``await`` keyword is applied, it suspends the calling method and returns control back to its caller until the awaited task is complete.  This is what allows a UI to be responsive and a service to be elastic.
```
Should we use the term 'yield' somewhere?

richlander commented on docs/async/async-csharp-vb.rst:
```diff
@@ -0,0 +1,199 @@
+Example (C#)
+------------
+
+The following example shows how to write basic async code for both a client app and a web service.  The code, in both cases, will count the number of times "Microsoft" appears in the HTML of "microsoft.com".
```
You might say "appears in the HTML response of http://microsoft.com."

richlander commented on docs/async/async-csharp-vb.rst:
```diff
@@ -0,0 +1,199 @@
+	private async void SeeTheMicrosofts_Click(object sender, RoutedEventArgs e)
+	{
+	    // Capture the task handle here so we can await the background task later.
+	    var getMicrosoftDotComHtmlTask = _httpClient.GetStringAsync("microsoft.com");
```
Does that actually work, without the "http://"? Even if it does, I suspect it isn't best practice.

richlander commented on docs/async/async-csharp-vb.rst:
```diff
@@ -0,0 +1,199 @@
+
+.. code-block:: c#
+
+	private readonly HttpClient _httpClient = new HttpClient();
```
This is a fine example. With our focus on open source and broadening our engagement with the community, I'm wondering if we want to have such a corporate heavy example. Could you modify to point the .NET Foundation site instead?

cartermp replied:
Will do.  It'll now count the number of ".NET"s in the HTML.

richlander commented on docs/async/async-csharp-vb.rst:
```diff
@@ -0,0 +1,199 @@
+	private readonly HttpClient _httpClient = new HttpClient();
+
+	[HttpGet]
+	public async Task<int> MicrosoftCount()
```
How about "GetMicrosoftCountAsync"?

richlander commented on docs/async/async-csharp-vb.rst:
```diff
@@ -0,0 +1,199 @@
+More on Task and Task<T>
+------------------------
+
+As mentioned before, Tasks are constructs used to represent operations which you will need to wait on to complete.  Although Tasks are general, the important context here is using them to represent an I/O operation.
```
> the important context here is using them to represent an I/O operation

Why is that?

richlander commented on docs/async/async-csharp-vb.rst:
```diff
@@ -0,0 +1,199 @@
+* ``Task`` represents a single operation which does not return a value.
+* ``Task<T>`` represents a single operation which returns a value of type ``T``.
+
+It's important to reason about Tasks as abstractions of work to be done rather than threads, because in the context of asynchronous I/O they aren't threads at all!  Asynchronous .NET library calls essentially delegate work to the operating system, which is naturally asynchronous when I/O is performed.  Unless you write code using a Task to abstract the notion of a new thread (such as the ``Task.R ...
```
> because in the context of asynchronous I/O they aren't threads at all

They are not ever threads.

Try to avoid the word "essentially" when it doesn't really add anything. 

Lots of .NET APIs delegate work to the OS. In this case, the delegation is to a low-level file system driver that calls back when there is work to do. My understanding is that the file system has both sync and async modes. Assuming thats true, we should avoid the word "naturally".

The last sentence is fine. It might be improved by saying "Tasks are always executed (in terms of CPU) on the current thread, unless a separate thread is requested by using the Task.Run method."

cartermp replied:
> They are not ever threads.

`Task.Run` specifically does create a new thread, and is a recommended pattern to follow when you have CPU-bound work.  In that particular case, the task handle returned is representing a new thread.  This is why I'm stating that it's in the context of I/O.

> Lots of .NET APIs delegate work to the OS. In this case, the delegation is to a low-level file system driver that calls back when there is work to do. My understanding is that the file system has both sync and async modes. Assuming thats true, we should avoid the word "naturally".

I don't know about Windows internals that much, but IIRC in my operating systems class I read that internally, I/O requests are asynchronous and drivers are not allowed to block on I/O requests (unless the request can be processed *immediately* which is rarely the case).  On Linux that convention is followed as well, even if the code making the POSIX call is synchronous.  Especially if Linux is going to be a huge target system to focus on in these docs, it's very much a world where async, signals, and interrupts are the norm.

richlander commented on docs/async/async-csharp-vb.rst:
```diff
@@ -0,0 +1,199 @@
+
+It's important to reason about Tasks as abstractions of work to be done rather than threads, because in the context of asynchronous I/O they aren't threads at all!  Asynchronous .NET library calls essentially delegate work to the operating system, which is naturally asynchronous when I/O is performed.  Unless you write code using a Task to abstract the notion of a new thread (such as the ``Task.R ...
+
+Tasks are awaitable, meaning that the application of the ``await`` keyword will allow your system to wait for the Task to complete without blocking the executing thread.  If you're using ``Task<T>``, the ``await`` keyword will additionally "unwrap" the value returned.
```
Is "wait" the correct term? Does the system wait?

richlander commented on docs/async/async-csharp-vb.rst:
```diff
@@ -0,0 +1,199 @@
+
+* **You should add "Async" as the suffix of every async method name you write.**
+
+This is the convention used in .NET to more-easily differentiate synchronous and asynchronous methods.  It's better to be explicit and follow that patter here!  Note that certain methods which aren't explicitly called by your code (such as event handlers or web controller methods) don't really apply.  Because these are not explicitly called by your code, being explicit about them isn't important.
```
"pattern"

richlander commented on docs/async/async-csharp-vb.rst:
```diff
@@ -0,0 +1,199 @@
+
+* ``await`` **is what will ultimately make your method asynchronous.**
+
+Failing to apply the ``await`` operator inside your async method will make it run synchronously and not wait for any Tasks it calls to finish!  Application of ``await`` waits for a Task you called to complete and suspends your async method, giving back control to the method which called it.  This allows that calling method to perform other work on the main thread of execution rather than forcing  ...
```
This seems to duplicate content earlier in the document.

cartermp committed:
Merge remote-tracking branch 'refs/remotes/dotnet/master' into async-section

Conflicts:
	docs/concepts/primer.rst
	docs/index.rst

cartermp committed:
Add timeline and flesh out what's going on in an async call

richlander commented on docs/async/async-csharp-vb.rst:
```diff
@@ -0,0 +1,213 @@
+==========================
+By `Phillip Carter`_
+
+Async programming in C# and VB share a language-level asynchronous programming model which allows for easily writing asynchronous code without having to juggle callbacks or conform to a library which supports asynchronicity.
```
We usually just say "asynchrony". It's shorter.

richlander commented on docs/async/async-csharp-vb.rst:
```diff
@@ -0,0 +1,213 @@
+
+Async programming in C# and VB share a language-level asynchronous programming model which allows for easily writing asynchronous code without having to juggle callbacks or conform to a library which supports asynchronicity.
+
+The core of it all are the ``Task`` and ``Task<T>`` objects, which model asynchronous operations.  The async and await keywords provide a natural and synchronous developer experience for interacting with tasks.  The result is the ability to write powerful code which cleanly expresses intent.  There are other ways to approach this than ``async`` and ``await``, outlined in the `Task-based Asynchron ...
```
"The core of it all" is a bit vague. You might say "The core of async" ???

You could say types instead of objects in the first sentence to make it super clear that you are talking about the definition. You are talking about modeling, so it kinda aligns. This one is more subtle, since there isn't really any loss of comprehension as currently written.

richlander commented on docs/async/async-csharp-vb.rst:
```diff
@@ -0,0 +1,213 @@
+
+The core of it all are the ``Task`` and ``Task<T>`` objects, which model asynchronous operations.  The async and await keywords provide a natural and synchronous developer experience for interacting with tasks.  The result is the ability to write powerful code which cleanly expresses intent.  There are other ways to approach this than ``async`` and ``await``, outlined in the `Task-based Asynchron ...
+
+For example, say you want to download some data from a web service when a button is pressed, but don't want to hang up the UI thread.  It can be accomplished simply like this:
```
Just say "For example, you may need to ...". The use of "say" seems unnecessary and colloquial.

"hang up" should be "block".

richlander commented on docs/async/async-csharp-vb.rst:
```diff
@@ -0,0 +1,213 @@
+	    DoStuff(stringData);
+	};
+
+And that's it!  The code expresses the intent (downloading some data asynchronously) without getting bogged down in interacting with Task objects.
```
Here, I don't mind "bogged down". It's colloquial, too, but I think it does a good job of communicating the meaning.

richlander commented on docs/async/async-csharp-vb.rst:
```diff
@@ -0,0 +1,213 @@
+	    var getDotNetFoundationHtmlTask = _httpClient.GetStringAsync("http://dotnetfoundation.com");
+
+	    // Any other work on the UI thread can be done here, such as enabling a Progress Bar.
+	    NetworkProgressBar.IsEnabled = true;
```
You might add that it is important to do this before you await the task to ensure that user gets the progress bar experience before execution is yielded.

richlander commented on docs/async/async-csharp-vb.rst:
```diff
@@ -0,0 +1,213 @@
+	[Route("DotNetCount")]
+	public async Task<int> GetDotNetCountAsync()
+	{
+	    // Suspends GetDotNetCountAsync() to allow the caller (the web server) to accept another request,
```
This comment line is almost going off the screen, at least in the GH rendering.

richlander commented on docs/async/async-csharp-vb.rst:
```diff
@@ -0,0 +1,213 @@
+	private async void SeeTheDotNets_Click(object sender, RoutedEventArgs e)
+	{
+	    // Capture the task handle here so we can await the background task later.
+	    var getDotNetFoundationHtmlTask = _httpClient.GetStringAsync("http://dotnetfoundation.com");
```
Here you use "http" and ".com". You use "https" and ".org" below. The second case looks better, assuming that the site supports https.

richlander commented on docs/async/async-csharp-vb.rst:
```diff
@@ -0,0 +1,213 @@
+	Private Async Sub SeeTheDotNets_Click(sender As Object, e As RoutedEventArgs)
+
+	    ' Capture the task handle here so we can await it later.
+	    Dim getDotNetFoundationHtmlTask As Task(Of String) = _httpClient.GetStringAsync("http://dotnetfoundation.org")
```
This line is going off the screen in the GH rendering.

richlander commented on docs/async/async-csharp-vb.rst:
```diff
@@ -0,0 +1,213 @@
+More on Task and Task<T>
+------------------------
+
+As mentioned before, Tasks are constructs used to represent operations which you will need to wait on to complete.
```
I wonder what the correct terminology is. Do you "wait" on tasks to complete?

richlander commented on docs/async/async-csharp-vb.rst:
```diff
@@ -0,0 +1,213 @@
+* ``Task`` represents a single operation which does not return a value.
+* ``Task<T>`` represents a single operation which returns a value of type ``T``.
+
+It's important to reason about Tasks of abstractions of work happening in the background, and *not* dedicated threads on the thread pool, because in the context of I/O-bound code, they aren't new threads at all!
```
> Tasks of abstractions of work

That's a bit vague.

"they" is a bit vague. Tasks are not thread ever. I think you want to focus on "using" threads, not "being" threads.

richlander commented on docs/async/async-csharp-vb.rst:
```diff
@@ -0,0 +1,213 @@
+
+Here's a 10,000 foot view of what happens with a typical async call:
+
+The call (such as ``GetStringAsync`` from ``HttpClient``) makes its way through the .NET libraries until it reaches a system interop call (such as ``P/Invoke`` on Windows).  This eventually makes the proper System API call (such as ``write`` to a socket file descriptor on Linux).  That System API call is then dealt with in the kernel, where the I/O request is sent to the proper subsystem.  Althou ...
```
> This will be converted into a ``Task`` or ``Task<T>`` by the runtime, sending that handle to the executing method which invoked the task in the first place.

How about:

This will be converted into a ``Task`` or ``Task<T>`` by the runtime and returned to the calling method.

richlander commented on docs/async/async-csharp-vb.rst:
```diff
@@ -0,0 +1,213 @@
+
+Tasks are awaitable, meaning that the use ``await`` will allow your application or service to perform useful work while the task is running by yielding control to its caller until the task is done.  If you're using ``Task<T>``, the ``await`` keyword will additionally "unwrap" the value returned when the Task is complete.
+
+Although the above may seem like a lot of work to be done, when measured in terms of execution time, it's miniscule compared to the time it takes to do the actual I/O work.  Although not at all precise, a potential timeline for such a call would look like this:
```
We usually like to use two precise terms: CPU time, wall-clock time. They are clear, whereas "execution time" can refer to either.

richlander commented on docs/async/async-csharp-vb.rst:
```diff
@@ -0,0 +1,213 @@
+
+0-1--------------------------------------------------------------------------------------------------------------------------------------------------2-3
+
+* Time spent from points ``0`` to ``1`` is everything up until an async method yield control to its caller.
```
"yields"

richlander commented on docs/async/async-csharp-vb.rst:
```diff
@@ -0,0 +1,213 @@
+
+* Time spent from points ``1`` to ``2`` is the time spend on I/O.
+
+* Finally, time spent from ``2`` to ``3`` is passing control back to the async method, at which point it is executing again.
```
You might add "(and potentially a value)"

richlander commented on docs/async/async-csharp-vb.rst:
```diff
@@ -0,0 +1,213 @@
+
+* ``async void`` **should only be used for event handlers.**
+
+Why?  It's the only way to allow asynchronous event handlers work because events do not have return types (thus cannot make use of ``Task`` and ``Task<T>``).  Any other use of ``async void``does not follow the Task-based model and will bring about some issues, such as:
```
"Why?" isn't really needed. You don't use that for the first statement.

"handlers to work"

richlander commented on docs/async/async-csharp-vb.rst:
```diff
@@ -0,0 +1,213 @@
+
+* **Tread carefully when using async lambdas in LINQ expressions**
+
+Lambda expressions in LINQ use deferred execution, meaning code could end up executing at a time when you're not expecting it to.  The introduction of blocking tasks into this can easily result in a deadlock if not written correctly.  The nesting of asynchronous code like this can also make it more difficult to reason about the execution of the code.  Async and LINQ are powerful, but if they are  ...
```
"they can make things difficult" is vague. Can you make it more specific?

richlander commented on docs/async/async-csharp-vb.rst:
```diff
@@ -0,0 +1,213 @@
+
+* **Write code that awaits Tasks in a non-blocking manner**
+
+Blocking the current thread as a means to wait for a Task to complete can result in deadlocks, blocked context threads, and significantly more complex error-handling.  The following table should provide some guidance in how to deal with waiting for Tasks in a non-blocking way:
```
Do you mean "and required significantly more complex ..."

You can say: "The following table provides guidance on how to deal with ..." We know it provides the guidance. It's not a question of "should".

richlander commented on docs/async/async-csharp-vb.rst:
```diff
@@ -0,0 +1,213 @@
+
+Why?  It's the only way to allow asynchronous event handlers work because events do not have return types (thus cannot make use of ``Task`` and ``Task<T>``).  Any other use of ``async void``does not follow the Task-based model and will bring about some issues, such as:
+
+    (a) Exceptions thrown in an ``async void`` method can't be caught outside of that method.
```
Would these be better as bullets, since they are unordered?

richlander commented on docs/async/async-csharp-vb.rst:
```diff
@@ -0,0 +1,213 @@
+
+Don't depend on the state of global objects or the execution of certain methods.  Instead, depend only on the return values of methods.  Why?
+
+	(a) Code will be easier to reason about
```
Would these be better as bullets?

I typically end these points with a ".".

richlander commented on docs/async/async-csharp-vb.rst:
```diff
@@ -0,0 +1,213 @@
+	(e) Depending on return values makes coordinating async code simple
+	(f) (Bonus) it works really well with dependency injection
+
+For a specific goal, aim for complete or near-complete `Referential Transparency <https://en.wikipedia.org/wiki/Referential_transparency_(computer_science)>`_ in your code.  Doing so will result in an extremely predictable, testable, and maintainable codebase.
```
"For a specific goal" seems like an odd statement. I don't know how to think of that.

richlander commented on docs/async/async-fsharp.rst:
```diff
@@ -0,0 +1,186 @@
+=======================
+By `Phillip Carter`_
+
+Async programming in F# can be accomplished through a language-level programming model designed to be easy to use and look just like any other F# code.
```
It looks like some of my feedback from the C# doc can be applied here. I won't go line by line until you've had an opportunity to do that.

richlander commented on docs/async/async-overview.rst:
```diff
@@ -0,0 +1,29 @@
+* Client applications are expected to be always-on and always-connected, but also keep the UI free for the user to interact with.
+* Services are expected to handle spikes in traffic by gracefully scaling up and down.
+
+.NET provides the capability for apps and services to be responsive and elastic with easy-to-use, language-level asynchronous programming models in C#/VB and F#.
```
Here, I don't think you need to write C#/VB, just because their implementation is so similar. A CSV list is fine.

cartermp committed:
Merge remote-tracking branch 'refs/remotes/dotnet/master' into async-section

cartermp committed:
Address richlander comments and improve F# doc

cartermp committed:
Address richlander comments more

Changed some wording to be friendlier, added a comment, and cleaned up a
few mistakes.

richlander commented:
I'm happy to go with this now. Squash and merge?

cartermp commented:
I messed things up in my environment, preventing me from merging this in properly. Going to open a new PR with the squashed commits.
